src/hdmf/data_utils.py

```diff
@@ -125,6 +127,225 @@ def maxshape(self):
         raise NotImplementedError("maxshape not implemented for derived class")
 
 
+class GenericDataChunkIterator(AbstractDataChunkIterator):
+    """DataChunkIterator that lets the user specify chunk and buffer shapes."""
+
+    @docval(
+        dict(
+            name="chunk_mb",
+            type=(float, int),
+            doc="Size of the HDF5 chunk in megabytes. Recommended to be less than 1MB.",
+            default=None,
+        )
+    )
+    def _get_default_chunk_shape(self, **kwargs):
+        """Select chunk size less than the threshold of chunk_mb, keeping the dimensional ratios of the original data."""
+        chunk_mb = getargs('chunk_mb', kwargs)
+        assert chunk_mb > 0, f"chunk_mb ({chunk_mb}) must be greater than zero!"
+
+        n_dims = len(self.maxshape)
+        itemsize = self.dtype.itemsize
+        chunk_bytes = chunk_mb * 1e6
+        v = np.floor(np.array(self.maxshape) / np.min(self.maxshape))
+        prod_v = np.prod(v)
+        while prod_v * itemsize > chunk_bytes and prod_v != 1:
+            v_ind = v != 1
+            next_v = v[v_ind]
+            v[v_ind] = np.floor(next_v / np.min(next_v))
+            prod_v = np.prod(v)
+        k = np.floor((chunk_bytes / (prod_v * itemsize)) ** (1 / n_dims))
+        return tuple([min(int(x), self.maxshape[dim]) for dim, x in enumerate(k * v)])
+
+    @docval(
+        dict(
+           name="buffer_gb",
+           type=(float, int),
+           doc="Size of the data buffer in gigabytes. Recommended to be as much free RAM as safely available.",
+           default=None,
+        )
+    )
+    def _get_default_buffer_shape(self, **kwargs):
+        """
+        Select buffer size less than the threshold of buffer_gb, keeping the dimensional ratios of the original data.
+
+        Assumes the chunk_shape has already been set.
+        """
+        buffer_gb = getargs('buffer_gb', kwargs)
+        assert buffer_gb > 0, f"buffer_gb ({buffer_gb}) must be greater than zero!"
+        assert all(np.array(self.chunk_shape) > 0), f"Some dimensions of chunk_shape ({self.chunk_shape}) are less than zero!"
+
+        k = np.floor(
+            (buffer_gb * 1e9 / (np.prod(self.chunk_shape) * self.dtype.itemsize)) ** (1 / len(self.chunk_shape))
+        )
+        return tuple([
+            min(max(int(x), self.chunk_shape[j]), self.maxshape[j])
+            for j, x in enumerate(k * np.array(self.chunk_shape))
+        ])
+
+    __docval_init = (
```

Comment:
why define this outside the constructor?

Reply:
(a) Makes the constructor cleaner and easier to read,
(b) Leaves the door open for if an advanced user wants to override it to give better performance on their specific data. I designed the one-shot calculation around 'typical' data shapes we usually see in `ecephys`, but there may be edge cases where it isn't as performant and the user might for example want to apply an exponentialized scaling preference for one axis over another. Or if they wanted really tight bounds with the size arguments, they could add iteration (which we avoid here because for `typical` data shapes we work with the one-shot calculation gets it pretty close). This is also why I moved many of the constructor shaping assertions to affect both user-specified shapes and those generated by these functions.